### PR TITLE
Improvements encountered during RECIPE

### PR DESCRIPTION
**Pull request by BGR360**
UPDATE: Rebased to be on top of Ian's big threading changes.

List of changes:

- Fix some issues with Ian's big threading changes
- Kind of a hacky thing to allow jemalloc library ctor to use our posix runtime
- Fix pthread_self() behavior enough to please jemalloc
- Passthrough `prefetch` intrinsics so I can run RECIPE
- Much more robust address resolution for `executePersistentMemoryFlush`

### CODE REVIEW

File: lib/Core/AddressSpace.cpp [hunk 1]
```diff
@@ -276,6 +276,97 @@ bool AddressSpace::resolve(ExecutionState &state, TimingSolver *solver,
...
+                                ResolutionList &rl, unsigned maxResolutions,
+                                time::Span timeout) const {
+  TimerStatIncrementer timer(stats::resolveTime);
+
+  // Sanity check range
+  bool mustBeTrue;
+  if (!solver->mustBeTrue(state, UltExpr::create(begin, end),
+                          mustBeTrue)) {
+    assert(false && "FIXME: Unhandled solver failure");
+  }
```
Comment (BGR360) [inline lib/Core/AddressSpace.cpp:313]: oops, need to actually fail if not mustBeTrue

File: lib/Core/Executor.cpp [hunk 2]
```diff
@@ -3942,60 +3942,129 @@ void Executor::executePersistentMemoryFlush(ExecutionState &state,
...
+  ref<Expr> rangeEnd = AddExpr::create(rangeBegin,
+                                       memory->getCacheAlignmentExpr());
+
+  ResolutionList rl;
   solver->setTimeout(coreSolverTimeout);
-  if (!state.addressSpace.resolveOne(state, solver, address, op, success)) {
-    address = toConstant(state, address, "resolveOne failure");
-    success = state.addressSpace.resolveOne(cast<ConstantExpr>(address), op);
-  }
+  bool incomplete = state.addressSpace.resolveRange(state, solver,
```
Comment (iangneal) [inline lib/Core/Executor.cpp:3991]: What's wrong with `state.addressSpace.resolve(state, solver, address, rl)`?

File: lib/Core/MemoryManager.cpp [hunk 3]
```diff
@@ -191,8 +191,20 @@ size_t MemoryManager::getUsedDeterministicSize() {
...
+ref<Expr> MemoryManager::getCacheAlignmentExpr(Expr::Width width) const {
+  return ConstantExpr::create(getCacheAlignment(), width);
+}
+
 uint64_t MemoryManager::alignToCache(uint64_t addr) const {
-  return addr - (addr % cacheAlignment);
+  return addr & ~(getCacheAlignment() - 1);
+}
+
+ref<Expr> MemoryManager::alignToCache(ref<Expr> addr) const {
```
Comment (iangneal) [inline lib/Core/MemoryManager.cpp:202]: Why do we need to align stuff to cache lines before hitting a `PersistentState`?

File: nvmbugs/002_simple_symbolic/simple_symbolic.c [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#include <stdlib.h>
+#include <stdio.h>
```
Comment (iangneal) [inline nvmbugs/002_simple_symbolic/simple_symbolic.c:2]: This seems like a duplicate of `002_SymbolicOffsetTest`

File: lib/Core/Executor.cpp [hunk 5]
```diff
@@ -3942,60 +3942,129 @@ void Executor::executePersistentMemoryFlush(ExecutionState &state,
...
+
+  if (incomplete) {
+    terminateStateEarly(state, "Query timed out (resolve).");
+  } else if (rl.empty()) {
+    terminateStateOnError(state, "memory error: out of bound pointer", Ptr,
+                          NULL, getAddressInfo(state, address));
+    return;
+  }
+
+  auto doFlush = [&](ExecutionState &state, const MemoryObject *mo,
```
Comment (iangneal) [inline lib/Core/Executor.cpp]: This function is long enough that it would be more readable to separate this out into a different `Executor` method with a docstring and all that jazz.

File: lib/Core/Executor.cpp [hunk 6]
```diff
@@ -3942,60 +3942,129 @@ void Executor::executePersistentMemoryFlush(ExecutionState &state,
...
+                     PersistentState *ps, ref<Expr> offset)  {
+
+    // Check if offset already flushed.
+    ref<Expr> check = ps->getIsOffsetPersistedExpr(offset,
+                                                   true /* allow pending */);
+
+    StatePair isAlreadyPersisted = fork(state, check, true);
```
Comment (iangneal) [inline lib/Core/Executor.cpp]: Hold on, I thought we were avoiding forking. Why not just check for if it's persisted, then if so, emit an error, else flush?

File: lib/Core/AddressSpace.h [hunk 7]
```diff
@@ -100,6 +111,19 @@ namespace klee {
                  unsigned maxResolutions=0,
                  time::Span timeout=time::Span()) const;
 
+    /// Resolve range `[begin, end)` to a list of `ObjectPairs` that
+    /// could overlap that range. If `maxResolutions` is non-zero then
+    /// no more than that many pairs will be returned.
+    ///
+    /// \return true iff the resolution is incomplete (`maxResolutions`
+    /// is non-zero and it was reached, or a query timed out).
+    bool resolveRange(ExecutionState &state,
```
Comment (iangneal) [inline lib/Core/AddressSpace.h:123]: Why is this necessary? It seems when you invoke this, the range is within a cacheline or something. It's not clear to me why you need this.

**Review by iangneal — CHANGES_REQUESTED** [review]

File: nvmbugs/002_simple_symbolic/simple_symbolic.c (hunk 4 above)
Comment (BGR360) [reply nvmbugs/002_simple_symbolic/simple_symbolic.c:2]: hm, oops, might be. I only looked at the other 002

File: lib/Core/Executor.cpp (hunk 6 above)
Comment (BGR360) [reply lib/Core/Executor.cpp]: This is how it was, I didn't add forking.

File: lib/Core/AddressSpace.h (hunk 7 above)
Comment (BGR360) [reply lib/Core/AddressSpace.h:123]: I explained this in the commit message, but I should add a comment too:

Some programs round their pointers to be cache-aligned
when flushing modifications to pmem. This means that it's
possible for the address passed to clwb/clflush to *not*
resolve to any known MemoryObject:
```
    Known MemoryObject: 0x10008, size 8
    Program calls clwb(0x10000)
```

Solution: rather than searching for MemoryObjects that contain a single
pointer, resolve all MemoryObjects that overlap any part of the cache
line being flushed. New function `AddressSpace::resolveRange` added to
help achieve this. Also led to more complicated executeFlush logic.

File: lib/Core/Executor.cpp (hunk 6 above)
Comment (BGR360) [reply lib/Core/Executor.cpp]: But, I think you're probably right.

File: lib/Core/AddressSpace.h (hunk 7 above)
Comment (iangneal) [reply lib/Core/AddressSpace.h:123]: Gotcha. Yeah put this in a comment, it's easier than git logging for documentation.

It actually seems like we should be using this for stores to persistent memory as well (especially for non-temporals), as you could construct a single store instruction that could dirty more than one cache line (i.e., a vector write of two 64-bit ints could store one at the end of one cache line and the next at the beginning of another cache line.)

Comment (BGR360) [reply lib/Core/AddressSpace.h:123]: Address resolution already works properly for loads and stores.

Comment (iangneal) [reply lib/Core/AddressSpace.h:123]: Oh right, because it ends up calling write8 a bunch.

File: nvmbugs/002_simple_symbolic/simple_symbolic.c (hunk 4 above)
Comment (BGR360) [reply nvmbugs/002_simple_symbolic/simple_symbolic.c:2]: I remember, it is a duplicate, but not using mmap, just using klee_mark_persistent on a stack buffer.

File: lib/Core/AddressSpace.cpp [hunk 8]
```diff
@@ -276,6 +276,98 @@ bool AddressSpace::resolve(ExecutionState &state, TimingSolver *solver,
...
+  bool mustBeTrue;
+  if (!solver->mustBeTrue(state, inRange, mustBeTrue))
+    return 1;
+
+  if (mustBeTrue) {
+    rl.push_back(op);
+    return 0;
+  }
+
+  return 2;
```
Comment (iangneal) [inline lib/Core/AddressSpace.cpp:299]: Do you ever use this 2 value anywhere? It seems like you only ever check if ==1

**Review by iangneal — CHANGES_REQUESTED** [review]
